### ISSUE TITLE
Changes in Tensor relevant to oneDNN

### ISSUE THREAD

**Issue by jczaja** [PaddlePaddle/Paddle#43525]
### 问题描述 Please describe your issue

Hi,

I would like to explain what changes we are working on to implement in *Tensor class and oneDNN kernels. Shortly speaking Tensor was holding oneDNN memory format (4 bytes) and now it is holding oneDNN memory descriptor (400 bytes).

Main reason for that is oneDNN is operating using memory descriptor as an element used for creation of memory objects.
In the past (2+ years ago) it was memory format. oneDNN migrated to using memory descriptor so we in paddle should do this as well otherwise we need to use lots of hacks to have it all working and not every operation can be implemented.

To do this transformation we need:
1) Modify Tensor (replace format with memory desc)
2) Modify Data Transformation layer 
3) Modify operators' oneDNN kernels 

Here is the picture of situation as it look like for operator's oneDNN kernel using **memory format** (old way):

![before](https://user-images.githubusercontent.com/15085062/173623108-ef348851-7b64-4e19-bbfb-c8194ff04f1e.png)

And here is a simplified picture of situation for the same operator's oneDNN kernel  using **memory descriptor** (new way):

![after](https://user-images.githubusercontent.com/15085062/173623102-d8a2a776-fca4-4aff-b85c-ac189dbb9f24.png)

So on the first picture you can notice two problems:
1) Input Tensor is holding **memory format** and then inside kernel we need to recreate **memory descriptor** . This is minor problem but makes our kernel doing additional creation of structures
2) To put into output Tensor **memory format** of output memory object we need to guess  **memory format** from **memory descriptor**. This  is not always possible and each time new layouts are added into oneDNN library then we need to update guessing code. So this is hacky and unreliable.

So introduced majority of changes related to this topic but still has some more kernels to modify . At the very last PR in this series we will remove from Tensor : set_layout()  and set_format() which are left now for compatibility.

To summarize, oneDNN evolved over the years to keep up we need to introduce presented changes.

Feel free to ask us questions on this

**Comment by yaomichael** [comment]
@Aganlengzi FYI, this is to explain the work for #42322

**Comment by chenwhql** [comment]
Thanks, we know the reason why mem_desc is introduced here, but this is still not a good design. In addition to the volume problem, the memory layout of DenseTensor in paddle compiled with MKLDNN is different from the DenseTensor in paddle compiled without MKLDNN. We will  try to adjust the design here, and discuss with you when we have an idea.

**Comment by Aganlengzi** [comment]
@yaomichael @jczaja hi, Thank you, I closed https://github.com/PaddlePaddle/Paddle/issues/42322, we can discuss it later here. And please notice that set_layout() is not only used by MKLDNN.

**Comment by yaomichael** [comment]
@chenwhql  @Aganlengzi  you are still working on a redesign around set_layout(), aren't you?

**Comment by Aganlengzi** [comment]
@yaomichael hi, I think this issue can be closed since `mem_desc` issue has been explained. Future work about oneDNN related changes in DenseTensor or other can be discussed and done later.
